osu.Game.Rulesets.Osu/Difficulty/Skills/Aim.cs

```diff
@@ -14,51 +16,115 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
+            var osuPrevObj = (OsuDifficultyHitObject)Previous[0];
+            var osuLastObj = (OsuDifficultyHitObject)Previous[1];
+
+            var currVector = Vector2.Divide(osuCurrObj.JumpVector, (float)osuCurrObj.StrainTime);
+            var prevVector = Vector2.Divide(osuPrevObj.JumpVector, (float)osuPrevObj.StrainTime);
 
-            double result = 0;
+            // Start with regular velocity.
+            double aimStrain = currVector.Length;
 
-            if (Previous.Count > 0)
+            if (Precision.AlmostEquals(osuCurrObj.StrainTime, osuPrevObj.StrainTime, 10)) // Rhythms are the same.
             {
-                var osuPrevious = (OsuDifficultyHitObject)Previous[0];
+                if (osuCurrObj.Angle != null)
+                {
+                    double angle = osuCurrObj.Angle.Value;
+
+                    // Rewarding angles, take the smaller velocity as base.
+                    double angleBonus = Math.Min(currVector.Length, prevVector.Length);
+
+                    double wideAngleBonus = calcWideAngleBonus(angle);
+                    double acuteAngleBonus = calcAcuteAngleBonus(angle);
 
-                if (osuCurrent.Angle != null && osuCurrent.Angle.Value > angle_bonus_begin)
+                    if (osuCurrObj.StrainTime > 100)
+                        acuteAngleBonus = 0;
+                    else
+                    {
+                        acuteAngleBonus *= Math.Min(2, Math.Pow((100 - osuCurrObj.StrainTime) / 15, 1.5));
+                        wideAngleBonus *= Math.Pow(osuCurrObj.StrainTime / 100, 6);
+                    }
+
+                    if (acuteAngleBonus > wideAngleBonus)
+                        angleBonus = Math.Min(angleBonus, 150 / osuCurrObj.StrainTime) * Math.Min(1, Math.Pow(Math.Min(osuCurrObj.JumpDistance, osuPrevObj.JumpDistance) / 150, 2));
+
+                    angleBonus *= Math.Max(acuteAngleBonus * acute_angle_multiplier, wideAngleBonus * wide_angle_multiplier);
+
+                    // add in angle velocity.
+                    aimStrain += angleBonus;
+                }
+
+                if (prevVector.Length > currVector.Length)
                 {
-                    const double scale = 90;
+                    double velChangeBonus = Math.Max(0, Math.Sqrt((prevVector.Length - currVector.Length) * currVector.Length) - currVector.Length) * Math.Min(1, osuCurrObj.JumpDistance / 100);
 
-                    var angleBonus = Math.Sqrt(
-                        Math.Max(osuPrevious.JumpDistance - scale, 0)
-                        * Math.Pow(Math.Sin(osuCurrent.Angle.Value - angle_bonus_begin), 2)
-                        * Math.Max(osuCurrent.JumpDistance - scale, 0));
-                    result = 1.4 * applyDiminishingExp(Math.Max(0, angleBonus)) / Math.Max(timing_threshold, osuPrevious.StrainTime);
+                    aimStrain += velChangeBonus * vel_change_multiplier * Math.Sqrt(100 / osuCurrObj.StrainTime);
```

Comment:
I think `velChangeBonus` needs some sort of angle bonus implemented, I would say velocity changes are a lot harder on wide angles.

Reply:
added this by adding (1 + wideAngleBonus) multiplier.